pulp_container/app/registry.py

```diff
@@ -117,7 +130,67 @@ async def get_tag(self, request):
                 pk__in=await sync_to_async(repository_version.get_content)(), name=tag_name
             )
         except ObjectDoesNotExist:
-            raise PathNotResolved(tag_name)
+            if distribution.remote:
+                remote = await distribution.remote.acast()
+
+                relative_url = "/v2/{name}/manifests/{tag}".format(
+                    name=remote.namespaced_upstream_name, tag=tag_name
+                )
+                tag_url = urljoin(remote.url, relative_url)
+                downloader = remote.get_in_memory_downloader(url=tag_url)
+                response = await downloader.run(extra_data={"headers": V2_ACCEPT_HEADERS})
+
+                set_guid(generate_guid())
+                task = await sync_to_async(dispatch)(
+                    download_image_data,
+                    exclusive_resources=[repository_version.repository],
+                    kwargs={
+                        "repository_pk": repository_version.repository.pk,
+                        "remote_pk": remote.pk,
+                        "tag_name": tag_name,
+                        "response_data": response.data,
+                    },
+                )
+
+                # waiting shortly for the task to be completed since a container client could
+                # request related content (i.e., manifests and blobs) and halt the pull operation
```

Comment:
i am not sure this is good practice during pull operation.. did you test if the client backs off if you raise 429? or it just fails?

Reply:
I tested it with raising 429 just now. It helped.

I was not raising anything (even 429) in the beginning, so I thought waiting might be a better approach because of:
```
DEBU[0003] Error pulling candidate localhost:5001/docker-cache/library/busybox:latest: copying system image from manifest list: determining manifest MIME type for docker://localhost:5001/docker-cache/library/busybox:latest: reading manifest sha256:023917ec6a886d0e8e15f28fb543515a5fcd8d938edb091e8147db4efed388ee in localhost:5001/docker-cache/library/busybox: StatusCode: 404, 404: Not Found 
```

Yes, another option is to tell the client to wait explicitly with 429. Podman handled this case flawlessly and waited 2 seconds:
```
DEBU[0000] GET http://localhost:5001/v2/docker-cache/library/busybox/manifests/latest 
DEBU[0001] Too many requests to http://localhost:5001/v2/docker-cache/library/busybox/manifests/latest: sleeping for 2.000000 seconds before next attempt 
DEBU[0003] GET http://localhost:5001/v2/docker-cache/library/busybox/manifests/latest 
```

But, should I start returning 429 in every case? What if it takes more than 2 seconds (podman default backoff) to finish the task? We may dispatch another task unnecessarily. This is hard to reproduce. I was not able to find any podman option to change this default.

Reply:
Yes, we can background the task, by returning 429.

Reply:
I see your point @mdellweg however I am a little bit uncomfortable returning 429 right away. It means on every first pull operation the client will be throttled.

Reply:
can this task be immediate?

Reply:
In other instances we raise Throttled, can you check what's the difference? I  know that with Throttled there is exponential back off

Reply:
Why do we need to wait for the result of the test is what I don't understand.
If we need to then immediate is probably the solution. An that would be immediate only and if it failed we return 429, but the task would have been canceled at that time.

Reply:
Even with immediate tasks, this might take 2 seconds, which is blocking in the end.

Reply:
We need to ensure that the next time podman requests blobs, they are already present in the repository, otherwise 404 is returned. That is why we need to wait a bit, but not too long. Raising 429 is a safe bet because we explicitly tell podman to wait.

Reply:
What we need is some sort of atomic operation that will ensure that manifest, blobs and repo version will exist by the time we respond to the client with a manifest. Otherwise the client risks to ask for the blob which we have not yet created/added to the repo and he will get a 404 in the middle of the pull operation.
we have another option. We do all the work here in the app, and the work would consist in creating blobs/manifest content, remote artifacts, content associations, assigned created content to pending content and then dispatch and forget(i think) add_and_remove (immediate) task that will actually create repo version.

Reply:
next during pull operation we'd need looking besides the available content in the repo version also in the pending_content for the case when task has not finished yet

Reply:
there is one caveat - we do not have pending_* on the regular repo, just on the push repo model

Reply:
yet another reason to merge those types...

Reply:
I implemented pending_* on a regulare repo before, when I was handling the pull-through workflow from api-app. So, I do not think this is a dealbreaker. I am more worried about downloading listed manifests from content-app now.